src/components/OrdersWidget/OrderRow.tsx

```diff
@@ -339,6 +350,17 @@ const OrderRow: React.FC<Props> = props => {
     fetchToken(order.sellTokenId, order.id, networkId, setSellToken, isPendingOrder).catch(onError)
   }, [isPendingOrder, networkId, order, setBuyToken, setSellToken])
 
+  useEffect(() => {
+    // if token somehow not in list
+    // as it was traded already we assume it to be a valid token
+    // and know to user
+    // and add it to USER_TOKEN_LIST automatically
+    addTokenToListIfNotAlready({ token: sellToken, networkId })
+  }, [networkId, sellToken])
+  useEffect(() => {
+    addTokenToListIfNotAlready({ token: buyToken, networkId })
+  }, [networkId, buyToken])
```

Comment:
Although the logic is simple, I think this is not where we should add it

Imo, adding a token would be more a "service/api" job, more than a "view/component" job

Probably we should move it to a more central piece. When we load the orders. Even the hook would seem a better option.

If I make a new page, that list orders in a different way, and then it doesn't use `OrderRow` we would have this there.

Reply:
Where would you ove it? `useOrders` hook? Orders reducer?

Reply:
Maybe the reducer seems more general? It reacts to a state change
what do you think?

Reply:
i would move it to a middleware, should that be added

Reply:
Probably a service might be best.
While building the `trades` service I found myself in a situation this can be useful.

Suppose the page is loaded directly into a `trades` view where the orders/tokenlist have not been loaded yet. There might be trades for tokens not present in the local list yet.

Reply:
I'm good, with the service too. Where wever is more central/natural

Reply:
> Suppose the page is loaded directly into a `trades` view where the orders/tokenlist have not been loaded yet

If trades include unlisted tokenId, then a similar or better yet compound side-effect(around global reducer)should be added

If we go the service way then it would either mean adding a sideeffect to orders fetching, and we currently just call `exchangeApi.getOrdersPaginated` without going through a service
And then another side-effect for trades fetching.
Or some bigger abstraction around any fetching that returns token ids.

I prefer to keep logic closer to where new tokenIds can appear -- in reducers, also because it only becomes relevant when we both display orders/trade and later user wants to see tokens. No reason to check for new unlisted tokens on every call to service.
Logic will become:
```js
case 'OVERWRITE_ORDERS':
case 'UPDATE_ORDERS':
case 'APPEND_ORDERS':
+ case 'APPEND_TRADES':
 // add new tokens
```

Reply:
I'm currently using the token ids before getting into the global state: https://github.com/gnosis/dex-react/pull/1012/files#diff-1027fd89b6327be60ccab6646b57d502R65

The hook is calling the service and updating the global state. 
I didn't stop yet to think this through, but one option might be to re-use part of your logic here.

Reply:
You will touch reducer afterwards, right? To persisting trades to global state, to not refresh every time

Reply:
Yes, but at that point, I should have the tokens populated already.
And the reducer is already implemented: https://github.com/gnosis/dex-react/pull/1012/files#diff-7ef769cc38af3906b033b24919b284b2R21